kb/2022-03-10/multiple_nics_vm_connectivity.md

```diff
@@ -0,0 +1,140 @@
+
+For example:
+```
+$ ip route show default
+default via 192.168.0.254 dev eth0 proto dhcp src 192.168.0.100 metric 100
+$
+```
+
+If the issue covered by this KB occurs, you could login VM via Web Console in Harvester Web UI and execute the following command in your abnormal VM:
+```bash
+ip route show default
+```
+
+In abnormal state, you will get one default route with incorrect gateway IP.
+For example:
+```
+default via <Incorrect Gateway IP> dev eth0 proto dhcp src <VM's IP> metric 100
+```
+
+## Why the connectivity issue occurs randomly in VM
+
+In most of time, we will enable cloud based VM with DHCP on all NICs and enable DHCP server with gateway IP. When DHCP client receives response and gets gateway IP, it will setup IP address and default route to the gateway IP, so users can access VM via SSH to the IP.
+
+Mostly Linux distribution will treat multiple NICs without any dependency and allow only one default route with the same distance by default. It will start multiple DHCP clients simultaneously. If we have two NICs with DHCP and both have gateway IP in DHCP response, two DHCP clients will race for setting default route to Linux network. According to different implementations of DHCP script, the former and latter will not always win to set up the default route. This will cause an unstable default route in every DHCP renewing process or machine boot up. VM's connection will be unstable if the default route is changing.
+
+## How to avoid random connectivity issue
+
+We could avoid this issue easily by keeping VM only one NICs to serve network. We keep only one IP and one gateway in your system.
+
+But not always we could use only one NIC to connect. For example, our infrasturcture will have storage network and service network. For security reason, storage network will be isolated from service network and have an independent subnet. In this situation, we must have two NICs to serve both service and storage network.
+
+We could take one solution from below to fit your requirement and security policy.
+
+### Disable DHCP, always setup Static IP on secondary NIC
+
+The issue is caused by race condition of two DHCP clients. We could disable DHCP and always setup static IP in your VM setting. Or you could only set secondary NIC to static IP and keep primary one with DHCP enable.
+
+1. edit `/etc/sysconfig/network/ifcfg-eth0` as primay NIC
+
+```
+BOOTPROTO='static'
+IPADDR='192.168.0.100'
+NETMASK='255.255.255.0'
+```
+
+Or setup `eth0` with DHCP
+
+```
+BOOTPROTO='dhcp'
+DHCLIENT_SET_DEFAULT_ROUTE='yes'
+```
+
+
+2. edit route table `/etc/sysconfig/network/ifroute-eth0` (skipping if you used DHCP in eth0)
+
+Do not put other default route for your secondary NIC
+
+```
+# Destination  Dummy/Gateway  Netmask  Interface
+default        192.168.0.254  -        eth0
+```
+
+3. setup secondary NIC, edit `/etc/sysconfig/network/ifcfg-eth1`
+
+```
+BOOTPROTO='static'
+IPADDR='10.0.0.100'
+NETMASK='255.255.255.0'
+```
+
+#### Cloud-Init config
```

Comment:
Can we set the NIC names to the same as above or have a short introduction to the cloud-init file (when/where to use)

Reply:
enp1s0 or enp2s0 will be real-world naming in VM.
but the naming in Linux VM will not be always the same.
So I'm not sure.

if you prefer to replace with eth0/eth1 naming rule, I will follow your suggestion

Reply:
that's what I thought, maybe just an intro to the cloud-init might be needed then (instead of renaming the interfaces).

Reply:
got it
let me change it to eth0/eth1